### PR TITLE
Refactoring to simplify development of Rematch2.

### PR DESCRIPTION
**Pull request by gafter**
* Significantly improved pretty-printing of the state machine for debugging purposes.
* Introduced a naming scheme for temp variables so we can tell when they represent phi functions, which cannot be used across patterns.
* Added macros for testing and diagnosing optimization opportunities in the state machine.  They work just like the normal @match2 macros but they display the state machine and/or count the number of states.

The short form of a printed state machine looks like this
```
State Machine: (5 states) input «input_value»
State 1 TEST «input_value» isa Main.Foo ELSE: State 5 («label_0»)
State 2 FETCH «input_value.y» := «input_value».y
State 3 TEST «input_value.y» == 2 ELSE: State 5 («label_0»)
State 4 MATCH 1 with value 1
State 5 («label_0») FAIL (throw)((Rematch2.MatchFailure)(«input_value»))
end # of state machine
```

Which was produced by running this code:
```
struct Foo
    x
    y
end
function main()
    @Assert (Rematch2.@match2_dump_states some_value begin
        Foo(x, 2) => 1
    end) == 5
end
main()
```

These changes were originally tangled up with #23.  This should be merged after #23 is merged.

### CODE REVIEW

File: src/BindPattern.jl [hunk 1]
```diff
@@ -49,22 +50,45 @@ end
...
 end
-function gentemp(p::BoundFetchLengthPattern)
-    gensym(string("length(", simple_name(p.input), ")"))
+struct FrenchName; s::Symbol; end
+Base.show(io::IO, x::FrenchName) = print(io, pretty_name(x.s))
+function pretty(io::IO, expr::Expr)
+    b = MacroTools.prewalk(MacroTools.rmlines, expr)
+    c = MacroTools.prewalk(MacroTools.unblock, b)
+    print(io, MacroTools.postwalk(c) do var
+        (var isa Symbol) ? FrenchName(var) : var
```
Comment (mcmcgrath13) [inline src/BindPattern.jl:90]: it looks like `pretty(io::IO, s::Symbol)` ends up in the same place as `FrenchName` - what does `FrenchName` get us?

File: README.md [hunk 2]
```diff
@@ -1,9 +1,9 @@
 # Rematch2 - Pattern-matching for Julia
 
-[![Stable](https://img.shields.io/badge/docs-stable-blue.svg)](https://gafter.github.io/Rematch2.jl/stable/)
-[![Dev](https://img.shields.io/badge/docs-dev-blue.svg)](https://gafter.github.io/Rematch2.jl/dev/)
-[![Build Status](https://github.com/gafter/Rematch2.jl/actions/workflows/CI.yml/badge.svg?branch=main)](https://github.com/gafter/Rematch2.jl/actions/workflows/CI.yml?query=branch%3Amain)
-[![Coverage](https://codecov.io/gh/gafter/Rematch2.jl/branch/main/graph/badge.svg)](https://codecov.io/gh/gafter/Rematch2.jl)
+[![Stable](https://img.shields.io/badge/docs-stable-blue.svg)](https://JuliaServices.github.io/Rematch2.jl/stable/)
+[![Dev](https://img.shields.io/badge/docs-dev-blue.svg)](https://JuliaServices.github.io/Rematch2.jl/dev/)
+[![Build Status](https://github.com/JuliaServices/Rematch2.jl/actions/workflows/CI.yml/badge.svg?branch=main)](https://github.com/JuliaServices/Rematch2.jl/actions/workflows/CI.yml?query=branch%3Amain)
```
Comment (mcmcgrath13) [inline README.md:5]: could these repo move/license changes be a standalone PR?

File: src/BindPattern.jl [hunk 3]
```diff
@@ -93,7 +136,7 @@ end
 const phi_prefix = "saved_"
 is_phi(s::Symbol) = startswith(simple_name(s), phi_prefix)
 function get_temp(state::BinderState, p::BoundFetchPattern)
-    get!(() -> gentemp(p), state.assignments, p)
+    get!(state.assignments, p) do; gentemp(p); end
```
Comment (mcmcgrath13) [inline src/BindPattern.jl:139]:
```suggestion
    get!(gentemp, state.assignments, p)
```

File: src/BindPattern.jl [hunk 4]
```diff
@@ -280,15 +323,9 @@ function bind_pattern!(
...
-        ### TODO: make this more dry. Currently we repeat
-        ###    fetch_foo = Fetch...(...)
-        ###    foo_temp = get_temp(state, fetch_foo)
-        ###    push!(patterns, fetch_foo)
-
         # produce a check that the length of the input is sufficient
-        fetch_length = BoundFetchLengthPattern(location, source, input)
-        length_temp = get_temp(state, fetch_length)
-        push!(patterns, fetch_length)
+        length_temp = push_pattern!(patterns, state,
```
Comment (mcmcgrath13) [inline src/BindPattern.jl:327]: 🧹

File: src/BoundPattern.jl [hunk 5]
```diff
@@ -1,4 +1,6 @@
+using Base: ImmutableDict
 
+# Unfortunately, using a type alias instead of the written-out type tickles a Julia bug.
```
Comment (mcmcgrath13) [inline src/BoundPattern.jl:3]: has an issue been filed for this in the Julia repo? If so, could you link to it here?

File: src/BindPattern.jl (hunk 1 above)
Comment (gafter) [reply src/BindPattern.jl:90]: Since the outside call is `print`, we somehow need the symbols in the middle of the expression to be pretty-printed instead of printed.  So we build a new expression with the symbols replaced by their pretty form.

File: README.md (hunk 2 above)
Comment (gafter) [reply README.md]: As annoying as it is, I already have a tangled mess of PRs piled on top of one another, and I'd rather not try to reorganize them.

**Review by mcmcgrath13 — APPROVED** [review]
thanks!
